### DISCUSSION TITLE
Repository patterns with correct types

### DISCUSSION THREAD

**Discussion by chrislgrigg** [prisma/prisma#19700] [Q&A]
Hello!

I'm using a repository pattern to create simple abstractions around Prisma in order to keep dependencies under control. I want to expand it so it accepts an `include` object. The signatures now look like this:

```typescript
type ThingWithInclude<I extends Prisma.ThingInclude | undefined> = Thing & {
  [K in keyof I]: Prisma.ThingGetPayload<{ include: I }>;
};

export async function getThingById<I extends Prisma.ThingInclude>(
  { id, userId }: { id: string; userId?: string },
  include?: I,
)

// wraps getThingById and throws when null
export async function getThingByIdOrThrow<I extends Prisma.ThingInclude>(
  { id, userId }: { id: string; userId?: string },
  include?: I
)

// example usage

const maybeThing = (await getThingById({ id, userId }))?.id;
const definitelyThing = (await getThingByIdOrThrow({ id, userId })).id;
const definitelyThingWithFoo = (await getThingByIdOrThrow({ id, userId }, { foo: true })).foo;
```

This works but it it's quite ugly.  I have an implementation where I pass a `throw` argument into a single `getThingById` function but it requires explicit return types in a way that is unscalable. As is, I think the amount of custom typing is unscalable on its own and I think I can do better.

Can anyone tell me:

* Is there a mapped type that I can use to get pull the `Prisma.${T}Include` and `Prisma.${T}GetPayload` dynamically? TypeScript does not allow this example syntax because `Prisma` is a namespace.
* Is there a way of dynamically writing the repository function return type so it can infer the type correctly? I'm forced to explicitly set return type because the `throwError` argument controls whether the output is nullable, but I'd like to find a more elegant way of describing the output from Prisma.
* Is there a better way of describing the return from something like `findFirst` that may or may not include the `include` or `select` result?

Thank you! I'm finding that Prisma has a great developer experience when it comes to iterating on data models and keeping the basics simple and type safe. Building repositories has felt so far like it requires boilerplate and reinvention of the wheel. Any advice on doing this more elegantly would be greatly appreciated. I've found similar threads at https://github.com/prisma/prisma/discussions/10584 and those it links to but none seem to answer these questions.

**Comment by chrislgrigg** [comment]
I'm closing this. When I opened it just a few minutes ago, I started with a single `getThingById` function that allowed control over whether to throw using an argument. I split it out into separate functions for throw or not throw, which allows the functions to infer types correctly, which removes the need for the advanced types to understand what to do when `include` is passed, which really negated the whole question. 🙃 I updated a few times to reflect the changes but came to the conclusion that the question doesn't really make sense anymore. 

I think there might be a discussion here about how to build repositories without so much boilerplate. As is, I'm forced to do this for every repository function that requires complex inputs, and I feel like I'm creating helper functions that the Prisma client, in all its brilliance (seriously!) could probably do already. But... I'm going to limp along with this for now and come back with a new question when I've reached the limits of sanity.

**Reply by darr1s** [reply]
Hey @subvertallchris, would you be able to share an example?
